### PR TITLE
Add gorm.io/gorm trace instrumentation

### PR DESCRIPTION
**Pull request by kostyay**

### CODE REVIEW

**Comment by kostyay** [comment]
I'm not related to the official gorm team.
Did I place the package in the right directory path?

File: instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go
Comment (XSAM) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:55]: Should we align the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md) by changing the span name?

**Review by XSAM — COMMENTED** [review]
> Did I place the package in the right directory path?

Yeah, the place seems good.

-----

Would you like to move the example into the `example` folder? You can take the other instrumentation as an instance.
Also, please add this instrumentation into `./instrumentation/README.md`.

File: instrumentation/github.com/go-gorm/gorm/otelgorm/plugin_test.go
Comment (kostyay) [reply instrumentation/github.com/go-gorm/gorm/otelgorm/plugin_test.go:231]: Its required to skip tests in non "integration" environment

File: instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go
Comment (kostyay) [reply instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:55]: not sure i understand the question

**Comment by kostyay** [comment]
Fixed code review comments and rebase

File: instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go
Comment (XSAM) [reply instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:55]: > SQL spans SHOULD be named the following way, unless the statement is known to be of low cardinality: <db.operation> <db.name>.<db.sql.table>, provided that db.operation and db.sql.table are available.

For instance, a span name of a `SELECT` call might be `SELECT ShopDb.orders`; `ShopDb` is the database name, `orders` is the table.

> i saw that mongodb uses similar approach of mongodb.query

Good catch, I raised a [comment](https://github.com/open-telemetry/opentelemetry-go-contrib/issues/460#issuecomment-775812136) to address that.

Comment (kostyay) [reply instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:55]: The table name is not obvious or easily accessible in many use cases.
What if the select is performed with a join over several tables? I'm not sure that convention makes much sense.
In order to achieve the that specification, one must parse the SQL query and understanding the SQL syntax which is beyond the scope of what this plugin is meant to solve.

Comment (XSAM) [reply instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:55]: > The table name is not obvious or easily accessible in many use cases.

I found table can be accessed from `gorm.DB.Statement.Table`.

> What if the select is performed with a join over several tables? In order to achieve that specification, one must parse the SQL query and understanding the SQL syntax which is beyond the scope of what this plugin is meant to solve.

Then we should ignore the table, just `<db.operation> <db.name>`. Similarly, if the database name is not available, we should unset this field, just `<db.operation>`.

From the specification:

> If the operation is acting upon an anonymous table, or more than one table, this value MUST NOT be set.

> It is not recommended to attempt any client-side parsing of `db.statement`.

-----

I'd keep the span name in low cardinality but **distinct**, which can describe an "interesting class of spans".

Comment (kostyay) [reply instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:55]: `gorm.DB.Statement.Table` is not always available. 
I will do best effort resolving for that.

**Comment by kostyay** [comment]
@XSAM should be good now

File: instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go
Comment (XSAM) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:30]:
```suggestion
	dbTableKey     = label.Key("db.sql.table")
```
From Spec, we have `db.sql.table`, which records the name of the primary table that the operation is acting upon

File: instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go
Comment (XSAM) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:31]:
```suggestion
	dbRowsAffectedKey     = label.Key("db.rows_affected")
```

`db.count` seems unclear in its meaning.

File: instrumentation/github.com/go-gorm/gorm/otelgorm/doc.go
Comment (XSAM) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/doc.go:21]: We need an example for this instrumentation, you can take [this](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/github.com/bradfitz/gomemcache/memcache/otelmemcache/example) as a sample.

**Review by XSAM — COMMENTED** [review]
Looks good overall. All comments are for minor questions/asks.

**Comment by kostyay** [comment]
@XSAM fixed all comments

**Comment by XSAM** [comment]
@kostyay Thanks for the hard work. 

But I don't see examples. Did you forget to commit?

**Comment by kostyay** [comment]
> @kostyay Thanks for the hard work.
> 
> But I don't see examples. Did you forget to commit?

Didn't notice `example` folder is gitignored. Added it now :)

File: instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go
Comment (XSAM) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go:61]:
```suggestion

```

nit

File: instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go
Comment (XSAM) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go:94]:
```suggestion

```

File: instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go
Comment (XSAM) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go:98]:
```suggestion

```

**Review by XSAM — APPROVED** [review]
Thanks for your contribution!

**Comment by kostyay** [comment]
All is green :)

**Comment by kostyay** [comment]
How do we get this merged?
Do all the reviewers need to approve?

**Comment by XSAM** [comment]
> How do we get this merged?
> Do all the reviewers need to approve?

It has received two approvals from Approvers/Maintainers (at different companies).

**Comment by nicksherron** [comment]
Would love to see this merged! I know everyone's probably plenty busy but it looks like all this needs is one more approval no?

**Comment by kostyay** [comment]
@Aneurysm9 I've fixed what you asked and rebased over master.
Let me know if there are any additional requests to get this merged.

File: instrumentation/github.com/go-gorm/gorm/otelgorm/doc.go
Comment (Aneurysm9) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/doc.go:20]: This documentation doesn't look correct.

File: instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go
Comment (Aneurysm9) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/example/main.go:119]: Be sure to flush/shutdown the tracer provider at the end of this function to ensure the batched spans are sent to the exporter.

File: instrumentation/github.com/go-gorm/gorm/otelgorm/go.mod
Comment (Aneurysm9) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/go.mod:3]:
```suggestion
go 1.15
```

File: Makefile
Comment (Aneurysm9) [inline Makefile:102]:
```suggestion
	@if ./tools/should_build.sh gorm; then \
```

**Review by Aneurysm9 — COMMENTED** [review]
A few minor details to address, but generally this looks good to me.

**Comment by kostyay** [comment]
Fixed the comments

File: instrumentation/github.com/go-gorm/gorm/otelgorm/config.go
Comment (pellared) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/config.go:27]: This is not adhering to https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#option

See https://github.com/open-telemetry/opentelemetry-go-contrib/pull/857 how to easily address it

File: Makefile
Comment (pellared) [inline Makefile:101]: why `test-gorm` is implemented as "integration tests" ?
it does not require any Docker container as a dependency. could it not be a simple unit test like most of the other instrumentation?

Comment (kostyay) [reply Makefile:101]: i dont remember, i think it was failing to run on the ci due to sqlite dependencies

Comment (pellared) [reply Makefile:101]: The tests are not running on CI anyway because `INTEGRATION` env var is not set: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/508/checks?check_run_id=3055645335

File: instrumentation/github.com/go-gorm/gorm/otelgorm/plugin_test.go
Comment (pellared) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/plugin_test.go:227]: bug
```suggestion
			assert.Equal(tt, spans[0].SpanContext().TraceID(), spans[1].SpanContext().TraceID())
```

nit (maybe subjective): I like adding explicit assertion messages explaining why this assertion even exists like `should record spans under the same trace` - it just helps me understanding the tests quicker and also getting quicker insight when something fails

File: Makefile
Comment (pellared) [reply Makefile:101]: After fixing https://github.com/open-telemetry/opentelemetry-go-contrib/pull/508#discussion_r668631032 and `util.IntegrationShouldRun("test-gorm")` everything worked fine on my machine so I hope that it does not need to be an "integration test" 😉

File: instrumentation/github.com/go-gorm/gorm/otelgorm/plugin_test.go
Comment (kostyay) [reply instrumentation/github.com/go-gorm/gorm/otelgorm/plugin_test.go:227]: thats a good tip thanks

File: Makefile
Comment (kostyay) [reply Makefile]: i removed this, lets see if tests will pass

File: instrumentation/github.com/go-gorm/gorm/otelgorm/example/README.md
Comment (pellared) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/example/README.md:8]: nit:

```suggestion
## Running the example
```

otherwise `markdownlint` complains 😉

File: instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go
Comment (pellared) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:110]: It would be great if you add an option that will allow to not adding the `db.statement` attribute

Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#call-level-attributes

File: instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go
Comment (pellared) [inline instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go:102]: It is not visible in the tests if the query remains parametrized or it is transformed to a raw query.
Maybe worth adding just a single test with a parameterized query to show that `db.statement` contains a "raw query".

**Review by pellared — COMMENTED** [review]
Tomorrow, I plan to test how the instrumentation works for complex queries e.g. when [Common Table Expressions](https://www.sqlite.org/lang_with.html) are being used

**Comment by kostyay** [comment]
@pellared like i said integration tests fail in CI
`failed to initialize database, got error Binary was compiled with 'CGO_ENABLED=0', go-sqlite3 requires cgo to work.`
what do u want recommend doing?

**Comment by pellared** [comment]
> @pellared like i said integration tests fail in CI
> `failed to initialize database, got error Binary was compiled with 'CGO_ENABLED=0', go-sqlite3 requires cgo to work.`
> what do u want recommend doing?

Check what even sets `CGO_ENABLED=0`. 

I would probably also quickly check what happens if we add `export CGO_ENABLED = 1` in the `Makefile`.

**Comment by pellared** [comment]
@kostyay I have tired to fix the `failed to initialize database, got error Binary was compiled with 'CGO_ENABLED=0', go-sqlite3 requires cgo to work.` issue (I hope you are OK with it). The Go modules cache was in `.github/workflows/build_and_test.yml` was implemented in an inappropriate way... However, it did not help. Also I refactored the test code a little and added a possibility to manually trigger the `build_and_test` (it needs to be merged to work)... Feel free to remove my commits if you do not like them.

**Comment by pellared** [comment]
@kostyay  I just noticed that it is easy to reproduce locally! Just execute `make ci` and you will have the same problems. I hope it will help you to fix the problem faster.

**Comment by pellared** [comment]
@kostyay Fixed: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/508/commits/8e407f6219620b999b2ae20a7967fbd056f7c178 🎉

**Review by pellared — DISMISSED** [review]

**Comment by MrAlias** [comment]
Thanks for the contribution. According to our [new instrumentation policy](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation#new-instrumentation), we are no longer accepting new instrumentation to the project as we do not have the developer bandwidth to support it. If you are able, please consider maintaining this instrumentation in your own repository and listing it in the [the OpenTelemetry Registry](https://github.com/open-telemetry/opentelemetry.io#adding-a-project-to-the-opentelemetry-registry).

**Pull request closed without merging**
